### DISCUSSION TITLE
Switching from rocketunit to testbox

### DISCUSSION THREAD

**Discussion by michaeldiederich** [wheels-dev/wheels#1184] [General]
The current testing framework rocketunit has some cons and i think testbox has some pros, so I think about switching the testframework. How is the cfwheels plan?
Has someone a PR already? How would I integrate testbox into cfwheels?
What changes would I have to do, to get testbox integrated? I read the https://blog.adamcameron.me/2021/05/cfwheels-running-testbox-instead-of.html article and i have downloaded testbox/mockbox, but i am not right sure what is the best way to migrate both.

**Comment by adamcameron** [comment]
@michaeldiederich we took that approach for our real environment, and it works fine. IE: that exercise I blogged was not just an experimental one. I was writing TestBox tests this morning in fact (in a rare moment of still being able to write some code!).

I will say though that I am not a CFWheels expert, and the approach I document was what I needed to do to get things working. I can't vouch for it being a CFWheels-idiomatic way of doing it. I did have a coupla CFWheels experts offering me guidance though, so maybe it's OK. Or as OK as it needs to be.

CFWheels and its code does not lend itself to be tested due to fairly uncontrolled use of shared scopes, leaking variables & having a "kitchen sink" approach to OOP (which then leads to a kitchen-sink approach to application code too, it seems), but if you write yer code carefully/thoughtfully/OOP way and decouple it from CFWheels code as much as poss, it's easy to test with TestBox.

And TBH once we worked out a few things we need to set/mock/work-around for code that requires to be run in the context of the running framework, it's not _too_ arduous.

Sing out if you need any help eh? But here is not the best place to do so as I only occasionally look at it. The CFML Slack would be better. The CFWheels channel there got closed the other day  [frown], but there's a testing channel there. Or maybe if you `@` me here I'll get a ping to come look? Not sure.

Best of luck, and *good man* for putting effort into test automation!

-- 
Adam

**Reply by chapmandu** [reply]
Hey @michaeldiederich ,

There are definitely discussions around upgrading the test framework to testbox. 

I had not seen that article from @adamcameron regarding testbox and cfwheels. It's a great read, and he is pretty much spot on in his assessment of the difficulties in cleanly testing a wheels application.

I've not personally used testbox yet myself but would like to use it in my own cfwheels projects.

Rocketunit has served it's purpose and was a huge step forward for cfwheels when Tony implemented it some 10 years ago.. before that, there were no framework tests at all. 😱 

It's still a very clever bit of kit, but is best suited to unit tests only. 

Any contributions towards getting cfwheels running tests with testbox would be greatly appreciated.

**Comment by bpamiri** [comment]
@michaeldiederich ,

I'll also chime in to say that we are definitely looking at integration of TestBox into the next major version of CFWheels (most likely 3.0ish). I'd also like to look at WireBox and LogBox integrations. There are some fundamental core re-architecture that has to get worked out before we can do that though, or at least at the same time. 

So let us know how you get along on this, there is definitely interest in getting this accomplished so any experience you gain and can share will go a long way.

**Reply by adamcameron** [reply]
@bpamiri I would say that one of the problems with the way testing has been implemented in CFWheels thusfar *is* that you have "integrated" RocketUnit into CFWheels. It needs to run in the context of the framework for it to work. That's a fail.

Don't _integrate_ TestBox into CFWheels.

Maybe ship it as a dev dependency, 

Fix a few bugs that make it difficult for Wheels code to be called outside the context of the webroot which prevent Wheels code working at all with TestBox (see my blog article). But let TestBox be TestBox. Keep it separate from CFWheels.

(*) there is a bigger issue about how CFWheels code needs to be in a web-browsable dir at all - which it shouldn't - but that's for another thread.
